README.pod
=head1 NAME

HTTP::UA::Parser - Perl User Agent Parser

=for html
<a href="https://travis-ci.org/ua-parser/uap-perl"><img src="https://travis-ci.org/ua-parser/uap-perl.svg?branch=master"></a>

=head1 DESCRIPTION

Perl port of the ua-parser project - L<https://github.com/ua-parser>

=head1 SYNOPSIS

    use HTTP::UA::Parser;
    my $r = HTTP::UA::Parser->new();
    
    print $r->ua->toString();         # -> "Safari 5.0.1"
    print $r->ua->toVersionString();  # -> "5.0.1"
    print $r->ua->family;             # -> "Safari"
    print $r->ua->major;              # -> "5"
    print $r->ua->minor;              # -> "0"
    print $r->ua->patch;              # -> "1"
    
    print $r->os->toString();         # -> "iOS 5.1"
    print $r->os->toVersionString();  # -> "5.1"
    print $r->os->family              # -> "iOS"
    print $r->os->major;              # -> "5"
    print $r->os->minor;              # -> "1"
    print $r->os->patch;              # -> undef
    
    print $r->device->family;         # -> "iPhone"
    print $r->device->brand;          # -> "Apple"
    print $r->device->model;          # -> "iPhone"
    
=head1 Methods

=head2 new()

Accepts a user agent string to parse, leave empty to parse caller user agent.

=head2 parse()

Accepts a new user agent to parse

=head2 ua()

Parses browser part of the user agent

=head2 os()

Parsers operating system part of the user agent

=head2 device()

Parses device part of the user agent

=head1 Stringify Methods

Methods to print results as strings

=over 4

=item toString()

returns os / ua name

=item toVersionString()

returns full version number of os/browser

=item family()

returns family name of os/browser/device

=item major()

returns version's major part of os/browser

=item minor()

returns version's minor part of os/browser

=item patch()

returns versions patch part of os/browser

=item patchMinor()

returns version patch minor part of os/browser

=item brand()

returns brand name of device

=item model()

returns model name of device

=back
    
=head1 INSTALLATION

From CPAN shell simply type

    % perl -MCPAN -e 'install HTTP::UA::Parser'

Or from your local download, unpack and:

    % perl Makefile.PL
    % make && make test
    
Then install:

    % make install

=head1 COMMAND LINE

To update regexes.yaml file from command line
    
    % ua_parser -u

To parse some user agent from command line

    % ua_parser -p "some user agent"
    
Help usage

    % ua_parser -h

=head1 AUTHOR

Mamod A. Mehyar, E<lt>mamod.mehyar@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Mamod A. Mehyar & other contributors

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension HTTP::UA::Parser.

0.001  Sat Dec 23 23:25:21 2012
	- original version; created by h2xs 1.23 with options
		-X -n HTTP::UA::Parser

0.002  Tue Dec 25 23:25:21 2012
	- Adding command line options
	- When makefile script will install latest regexes.yaml file
	- Using Yaml::Tiny instead of Yaml::Any

0.003 Sun Jan 20 2013
	- Adding more tests using original ua-project tests yaml
	- Fixing "Other" devices detection

0.004 Wed Jun 5 2013
	- Check if port work with new regexes.yaml file
	- Even more tests
	- Fixing 0 and undef issue with UA parser
	- Adding makeParser functions

0.006 Fri Jul 10 2015
	- lots of changes thanks to @commenthol - https://github.com/commenthol
	- adding device brand and device model support by @commenthol
	- Code cleaning
	- Fix documentation


t/utils.pl
#!perl
use YAML::Tiny;
use LWP::UserAgent;

my $resources = 'https://raw.githubusercontent.com/ua-parser/uap-core/master/';
my $dir = "../uap-core/";

sub get_test_yaml {
    my $file = shift;
    my $content;
    
    if (-e $dir.$file) {
        $file = $dir.$file;
        open (my $FH, $file) or die "Can't open $file\n";
        $content = join("\n", <$FH>);
        close ($FH);
    }
    else {
        my $ua = LWP::UserAgent->new;
        my $req = HTTP::Request->new(GET => $resources . $file);
        my $res = $ua->request($req);
        die if !$res->is_success;
        $content = $res->content;
    }

    die if !$content;

    
    my $yaml = fix_yaml($content);
    return YAML::Tiny->read_string( $yaml )->[0]->{test_cases};
}

sub fix_yaml {
    my $yaml = shift;
    open(my $fh,'<', \$yaml );
    my $c;
    while(my $line = <$fh>){
        if ($line){
            $line =~ s/(\s*(?:family|major|minor|patch|patch_minor|brand|model):)\s?\n$/$1 ~\n/;
        }
        $c .= $line;
    }
    close $fh;
    return $c;
}

1;


Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
use FindBin qw($Bin);
use File::Copy;

My_WriteMakefile(
    NAME              => 'HTTP::UA::Parser',
    VERSION_FROM      => 'lib/HTTP/UA/Parser.pm',
    ABSTRACT_FROM  => 'lib/HTTP/UA/Parser.pm',
    EXE_FILES    => ['script/ua_parser'],
    PREREQ_PM         => {
        'YAML::Tiny' => 0
    },
    LICENSE        => 'perl',
    AUTHOR         => 'Mamod Mehyar <mamod.mehyar@gmail.com>',
    META_MERGE => {
        resources => {
            repository => 'git://github.com/ua-parser/uap-perl.git',
            bugtracker => 'https://github.com/ua-parser/uap-perl/issues',
        }
    }
);

sub My_WriteMakefile {
    my %params=@_;
    my $regexes = './uap-core/regexes.yaml';
    my $regexes2 = $Bin.'/regexes.yaml';
    my $dest = $Bin.'/lib/HTTP/UA/Parser';
    if (-e $regexes) {
        copy($regexes, $dest) or die "File cannot be copied.";
    } elsif (-e $regexes2) {
        copy($regexes2, $dest) or die "File cannot be copied.";
    } else {
        print "Can't locate regexes.yaml file\n";
        print "We will try to fetch it from a remote server\n";
        my $regexContent = getRegexes();
        open my $file,'>',$regexes2 or die "can't open file $!";
        print $file $regexContent;
        close $file;
        copy($regexes2, $dest) or die "File cannot be copied.";
        print "regexes.yaml downloaded successfully\n";
    }

    WriteMakefile(%params);
}

##also update in script/ua_parser if changed here
sub getRegexes {
    my $response;
    my $stream;
    my $url = 'https://raw.githubusercontent.com/ua-parser/uap-core/master/regexes.yaml';

    ## trying curl
    print "Trying curl\n";
    open $stream, "-|", "curl $url";
    while(<$stream>) { $response .= "$_" };

    ##trying wget
    if (!$response){
        print "Trying wget\n";
        open $stream, "-|", "wget $url";
        while(<$stream>) { $response .= "$_" };
    }

    ##trying lwp-request
    if (!$response){
        print "Trying lwp-request\n";
        open $stream, "-|", "lwp-request $url";
        while(<$stream>) { $response .= "$_" };
    }

    return $response if $response;
    print "Trying to fetch using LWP::UserAgent\n";
    eval "use LWP::UserAgent";
    if ($@){
        print "We couldn't locate LWP::UserAgent Module\n";
        print "LWP::UserAgent required to fetch regexes.yaml from server\n";
        print "Please install it or get regexes.yaml file manually from\n";
        print "$url\n";
        print "and place it in the root folder of this distro\n";
        print "then run Makefile.PL again\n";
        exit;
    }
    my $ua = LWP::UserAgent->new;
    $ua->timeout(5);
    $ua->env_proxy();
    $response = $ua->get($url);
    if ($response->is_success) {
        return $response->content;
    } else {
        print "Request aborted\n";
        exit;
    }
}


lib/HTTP/UA/Parser.pm
package HTTP::UA::Parser;
use strict;
use warnings;
use YAML::Tiny 'LoadFile';
our $VERSION = '0.006';
my ($REGEX, $PATH, $PARSER);
my $PACKAGE = __PACKAGE__;

sub new {
    my ($class, $op) = @_;
    my ($ua, $path);
    if (ref $op eq 'HASH'){
        $path = $op->{regex};
        $ua = $op->{ua};
    } else { $ua = $op; }
    
    if (!$REGEX){
        if ($path){
            $REGEX ||= LoadFile( $path );
        } else {
            $PATH = HTTP::UA::Parser::Utils::getPath();
            my $regFile;
            unless (-e ($regFile = $PATH.'/regexes.yaml')){
                die
                "Can't find regexes.yaml file\n".
                "you can download/update it using command line by typing\n".
                "    % ua_parser -u\n".
                "or simply download it from\n".
                "https://raw.githubusercontent.com/ua-parser/uap-core/master/regexes.yaml".
                "and include it as an option when construct new HTTP::UA::Parser class\n".
                "ex ->new({regex => '/full/path/to/regexes.yaml'})";
            }
            $REGEX = LoadFile( $regFile );
        }
    }
    
    my $self = {
        user_agent => $ua || $ENV{HTTP_USER_AGENT},
        path => $PATH
    };
    
    return bless($self,$class);
}

sub parse {
    my $self = shift;
    $self->{user_agent} = $_[0];
    $self->{os} = undef;
    $self->{ua} = undef;
    $self->{device} = undef;
    return $self;
}

sub os {
    my ($self) = @_;
    $self->{os} ||= HTTP::UA::Parser::OS->parse($self->{user_agent});
    return $self->{os};
}

sub ua {
    my ($self) = @_;
    $self->{ua} ||= HTTP::UA::Parser::UA->parse($self->{user_agent});
    return $self->{ua};
}

sub device {
    my ($self) = @_;
    $self->{device} ||= HTTP::UA::Parser::Device->parse($self->{user_agent});
    return $self->{device};
}

##=============================================================================
## UA Package
##=============================================================================
package HTTP::UA::Parser::UA;

sub new {HTTP::UA::Parser::Base::new(@_)}

sub parse {
    my $self = shift;
    my $ua = shift;
    my $parser = $PARSER->{ua} || ($PARSER->{ua} = $self->makeParser($REGEX->{user_agent_parsers}));
    return $parser->($ua);
}

sub makeParser {
    my $self = shift;
    my $regexes = shift;
    return HTTP::UA::Parser::Utils::makeParser($regexes, \&_makeParsers);
}

sub _makeParsers {
    
    my ($obj) = shift;
    my $regexp = $obj->{regex};
    my $famRep = $obj->{family_replacement};
    my $majorRep = $obj->{v1_replacement};
    my $minorRep = $obj->{v2_replacement};
    my $patchRep = $obj->{v3_replacement};
    my $qr = HTTP::UA::Parser::Utils::regex($regexp);
    
    my $parser = sub {
        my $str = shift;
        my @m = $str =~ $qr;
        if (!@m) { return undef; }
        my $family = defined $famRep ? HTTP::UA::Parser::Utils::replace($famRep,qr/\$1/,$m[0]) : $m[0];
        my $major = defined $majorRep ?  $majorRep : $m[1];
        my $minor = defined $minorRep ?  $minorRep : $m[2];
        my $patch = defined $patchRep ?  $patchRep : $m[3];
        return ($family, $major, $minor, $patch);
    };
    
    return $parser;
}

##=============================================================================
## OS Package
##=============================================================================
package HTTP::UA::Parser::OS;

sub new {HTTP::UA::Parser::Base::new(@_)}

sub parse {
    my $self = shift;
    my $ua = shift;
    my $parser = $PARSER->{os} || ($PARSER->{os} = $self->makeParser($REGEX->{os_parsers}));
    return $parser->($ua);
}

sub makeParser {
    my $self = shift;
    my $regexes = shift;
    return HTTP::UA::Parser::Utils::makeParser($regexes, \&_makeParsers);
}

sub _makeParsers {
    
    my ($obj) = shift;
    my $regexp = $obj->{regex};
    my $famRep = $obj->{os_replacement};
    my $majorRep = $obj->{os_v1_replacement};
    my $minorRep = $obj->{os_v2_replacement};
    my $patchRep = $obj->{os_v3_replacement};
    my $patchMinorRep = $obj->{os_v4_replacement};
    my $qr = HTTP::UA::Parser::Utils::regex($regexp);
    
    my $parser = sub {
        my $str = shift;
        my @m = $str =~ $qr;
        if (!@m) { return undef; }
        my $family = $famRep ? HTTP::UA::Parser::Utils::replace($famRep,qr/\$1/,$m[0]) : $m[0];
        my $major = defined $majorRep ? $majorRep : $m[1];
        my $minor = defined $minorRep ? $minorRep : $m[2];
        my $patch = defined $patchRep ? $patchRep : $m[3];
        my $patchMinor = defined $patchMinorRep ? $patchMinorRep : $m[4];
        return ($family, $major, $minor, $patch, $patchMinor);
    };
    
    return $parser;
}

##=============================================================================
## Device Package
##=============================================================================
package HTTP::UA::Parser::Device;

sub new {
    my $class = shift;
    my $self = {
        family => $_[0] || 'Other',
        brand => $_[1],
        model => $_[2]
    };
    return bless($self, 'HTTP::UA::Parser::Base');
}

sub toString {
    my $self = shift;
    return $self->family;
}

sub parse {
    my $self = shift;
    my $ua = shift;
    my $parser = $PARSER->{device} || ($PARSER->{device} = $self->makeParser($REGEX->{device_parsers}));
    return $parser->($ua);
}

sub makeParser {
    my $self = shift;
    my $regexes = shift;
    my $makeParser = shift || \&_makeParsers;
    my @parsers = map {
        $makeParser->($_);
    } @{$regexes};
  
    my $parser = sub {
        my $ua = shift;
        my @obj;
        foreach my $parser (@parsers){
            @obj = $parser->($ua);
            return HTTP::UA::Parser::Device->new(@obj) if $obj[0];
        }    
    
        HTTP::UA::Parser::Device->new();
    };
  
    return $parser;
}

sub _makeParsers {
    my ($obj) = shift;
    my $regexp = $obj->{regex};
    my $regexp_flag = $obj->{regex_flag};
    my $deviceRep = $obj->{device_replacement};
    my $brandRep = $obj->{brand_replacement};
    my $modelRep = $obj->{model_replacement};
    my $qr = HTTP::UA::Parser::Utils::regex($regexp, $regexp_flag);
    
    my $parser = sub {
        my $str = shift;
        my @m = $str =~ $qr;
        if (!@m) { return undef; }
        my $family = $deviceRep ? HTTP::UA::Parser::Utils::multiReplace($deviceRep, \@m) 
                                : ($m[0] eq "1" ? undef : $m[0]);

        my $brand  = $brandRep  ? HTTP::UA::Parser::Utils::multiReplace($brandRep, \@m)  
                                : undef;

        my $model  = $modelRep  ? HTTP::UA::Parser::Utils::multiReplace($modelRep, \@m)  
                                : ($m[0] eq "1" ? undef : $m[0]);
        
        return ($family, $brand, $model);
    };
    return $parser;
}

##=============================================================================
## Stringify Package
##=============================================================================
package HTTP::UA::Parser::Base;

sub new {
    my $class = shift;
    my $self = {
        family => $_[0] || 'Other',
        major => $_[1],
        minor => $_[2],
        patch => $_[3],
        patchMinor => $_[4]
    };
    return bless($self, __PACKAGE__ );
}

sub toVersionString {
    my $self  = shift;
    my $output = '';
    if (defined $self->{major}){
        $output .= $self->{major};
        if (defined $self->{minor}){
            $output .= '.' . $self->{minor};
            if (defined $self->{patch}) {
                if (HTTP::UA::Parser::Utils::startsWithDigit($self->{patch})) { $output .= '.'; }
                $output .= $self->{patch};
                if (defined $self->{patchMinor}) {
                    if (HTTP::UA::Parser::Utils::startsWithDigit($self->{patchMinor})) { $output .= '.'; }
                    $output .= $self->{patchMinor};
                }
            }
        }
    }
    return $output;
}

sub toString {
    my $self = shift;
    my $suffix = $self->toVersionString();
    if ($suffix){
        $suffix = ' ' . $suffix;
    }
    return $self->family . $suffix;
}

sub family     {   shift->{family}     }
sub major      {   shift->{major}      }
sub minor      {   shift->{minor}      }
sub patch      {   shift->{patch}      }
sub patchMinor {   shift->{patchMinor} }
sub brand      {   shift->{brand}      }
sub model      {   shift->{model}      }

##=============================================================================
## Utils Package
##=============================================================================
package HTTP::UA::Parser::Utils;

sub makeParser {
    my $regexes = shift;
    my $makeParser = shift;
    my @parsers = map {
        $makeParser->($_);
    } @{$regexes};
    
    my $parser = sub {
        my $ua = shift;
        my @obj;
        foreach my $parser (@parsers){
            @obj = $parser->($ua);
            return HTTP::UA::Parser::Base->new(@obj) if $obj[0];
        }
        
        HTTP::UA::Parser::Base->new();
    };
    
    return $parser;
}

sub replace {
    my ($stringToReplace,$expr,$replaceWith) = @_;
    $stringToReplace =~ s/$expr/$replaceWith/;
    return $stringToReplace;
}

sub multiReplace {
    my ($stringToReplace, $matches) = @_;
    for ($stringToReplace) {
        s{
            \$(\d)
        }{
            defined @{$matches}[$1-1] ? @{$matches}[$1-1] : '';
        }egx;
        s{^\s+|\s+$}{};
    }
    if ($stringToReplace eq '') {
        undef $stringToReplace;
    }
    return $stringToReplace;
}

# precompile regex
sub regex {
    my ($expr, $flag) = @_;
    $flag = $flag || '';
    if ($flag eq "i") {
        return qr{$expr}i;
    }
    return qr{$expr};
}

sub startsWithDigit {
    my $str = shift;
    return $str =~ /^\d/;
}

sub getPath {
    $PATH = $PACKAGE;
    $PATH =~ s/::/\//g;
    $PATH .= '.pm';
    $PATH = $INC{$PATH};
    $PATH =~ s/.pm$//;
    return $PATH;
}

1;

__END__
=head1 NAME

HTTP::UA::Parser - Perl User Agent Parser

=for html
<a href="https://travis-ci.org/ua-parser/uap-perl"><img src="https://travis-ci.org/ua-parser/uap-perl.svg?branch=master"></a>

=head1 DESCRIPTION

Perl port of the ua-parser project - L<https://github.com/ua-parser>

=head1 SYNOPSIS

    use HTTP::UA::Parser;
    my $r = HTTP::UA::Parser->new();
    
    print $r->ua->toString();         # -> "Safari 5.0.1"
    print $r->ua->toVersionString();  # -> "5.0.1"
    print $r->ua->family;             # -> "Safari"
    print $r->ua->major;              # -> "5"
    print $r->ua->minor;              # -> "0"
    print $r->ua->patch;              # -> "1"
    
    print $r->os->toString();         # -> "iOS 5.1"
    print $r->os->toVersionString();  # -> "5.1"
    print $r->os->family              # -> "iOS"
    print $r->os->major;              # -> "5"
    print $r->os->minor;              # -> "1"
    print $r->os->patch;              # -> undef
    
    print $r->device->family;         # -> "iPhone"
    print $r->device->brand;          # -> "Apple"
    print $r->device->model;          # -> "iPhone"
    
=head1 Methods

=head2 new()

Accepts a user agent string to parse, leave empty to parse caller user agent.

=head2 parse()

Accepts a new user agent to parse

=head2 ua()

Parses browser part of the user agent

=head2 os()

Parsers operating system part of the user agent

=head2 device()

Parses device part of the user agent

=head1 Stringify Methods

Methods to print results as strings

=over 4

=item toString()

returns os / ua name

=item toVersionString()

returns full version number of os/browser

=item family()

returns family name of os/browser/device

=item major()

returns version's major part of os/browser

=item minor()

returns version's minor part of os/browser

=item patch()

returns versions patch part of os/browser

=item patchMinor()

returns version patch minor part of os/browser

=item brand()

returns brand name of device

=item model()

returns model name of device

=back
    
=head1 INSTALLATION

From CPAN shell simply type

    % perl -MCPAN -e 'install HTTP::UA::Parser'

Or from your local download, unpack and:

    % perl Makefile.PL
    % make && make test
    
Then install:

    % make install

=head1 COMMAND LINE

To update regexes.yaml file from command line
    
    % ua_parser -u

To parse some user agent from command line

    % ua_parser -p "some user agent"
    
Help usage

    % ua_parser -h

=head1 AUTHOR

Mamod A. Mehyar, E<lt>mamod.mehyar@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Mamod A. Mehyar & other contributors

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.
